### PR TITLE
Draft: Add Travis CI support

### PR DESCRIPTION
**Pull request by Prabhav-Thali**
This refers to [issue-19688](https://github.com/eclipse/che/issues/19688). Adds Travis CI support to build multi-arch images. 

**Workflows migrated from GA to Travis CI:**

- PR check
- Code Coverage
- Docker build on push to main branch
- Unit tests 
- Release

In case of release workflow, we can trigger a manual build by using the trigger build option from Travis UI. After selecting branch, we can provide a custom config to override the config variables in .travis.yml. 
For eg:-
```
env: 
  global:
  - TAG=<version-to-be-released>
  - RECREATE_TAGS=true
  - NO_COMMIT=true
```

**Env variables to be set in Travis CI settings:**
DOCKER_PASSWORD 
DOCKER_USERNAME  
GITHUB_TOKEN  
MATTERMOST_WEBHOOK_URL  
QUAY_PASSWORD  
QUAY_USERNAME

Signed-off-by: Prabhav Thali <Prabhav.Thali1@ibm.com>

### CODE REVIEW

**Comment by Prabhav-Thali** [comment]
**cc**
@nickboldt

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+git:
+  depth: false
+
+services:
+  - docker
+
+env:  
+   global:
+   - TAG=nightly
+   - RECREATE_TAGS=false
```
Comment (nickboldt) [inline .travis.yml:18]: can this be changed to true via UI, like in a GH action?

Comment (nickboldt) [reply .travis.yml:18]: > we can trigger a manual build by using the trigger build option from Travis UI

ah, I see we can. :)

Can we also trigger this from an API? today we trigger all the GH actions for releasing Che from a single action: 
* https://github.com/eclipse/che-release/blob/master/make-release.sh#L134-L187
* https://github.com/eclipse/che-release/blob/master/.github/workflows/release-orchestrate-overall.yml

Is there a way to do a single "travis orchestrator" as well?

Comment (Prabhav-Thali) [reply .travis.yml:18]: I'll go through this and provide more details here

Comment (Prabhav-Thali) [reply .travis.yml:18]: > Can we also trigger this from an API?

Yes, we can trigger jobs in Travis using API. Ref: https://docs.travis-ci.com/user/triggering-builds/


> Is there a way to do a single "travis orchestrator" as well?

We can have a single Travis orchestrator in which GitHub API will be replaced with Travis API to trigger builds. I have tried the same for che-machine-exec and che-theia.

**Comment by nickboldt** [comment]
Notes from today's call w/ Florent and other Che devs:

* Travis limits us to 3 jobs in parallel (3 arches = 3 jobs); with GHA, our limit is 20 jobs in parallel
* Eric and possibly others agree that doing PR checks on multiple arches is vital, so that's a lot of potential parallel builds (more than 3)
* we might want to enable PR checks for x to be required and z/p to be optional, so as to not block merging commits
* Travis is not currently enabled for the eclipse-che org in GH (only for eclipse) - need to get that enabled before these PRs can be validated. For that we need to open a BZ to talk to Denis R (Eclipse WM) to approve
* But once it's approved, only need to pass tokens between travis and GHA to be able to trigger GHA->T or T->GHA

@cwsolee can you talk to your people at Travis to ask if we can get more capacity than 3 parallel jobs? If that's a hard limit, then we can't use travis for more than Z builds and it makes sense to keep everything else in the 20-build GHA pool (x, arm, power, etc.), and have a GH action delegate to Travis for only the s390x arch builds.

**Comment by cwsolee** [comment]
Travis change our concurrency limit too, it's 20 now and Travis were quite willing to increase more when needed. With both time limit & concurrency limit increase, we haven't hit any problem since then.

**Comment by Prabhav-Thali** [comment]
**Following changes are implemented:**

- TRAVIS_TAG is added such that each pushed tag is suffixed with "-travis". And tag for image on each arch will be in the format 7.x.x-travis-${arch}

- To handle the build failures that may occur only on non-amd64 arch, Added allow_failures and fast_finish=true labels in case of PR check. 

- In release workflow, removed execution of make-release.sh as it will conflict with GHA release. As tag is not pushed through Travis build, removed "check existing tag" step.

- Removed 2 workflows i.e. "code coverage : Test coverage report" and "license validation".

- Removed mattermost notification steps from Travis.yml.

- Added Travis API call in GHA workflows.

- Have added job for next build. Is nightly build job required to be added in Travis?


**PR check limitation on triggering PR-check jobs on Travis via API:**

- The Travis API token will be available to PRs raised by users with write access to that particular repo.

- Changes made to .travis.yml file in PRs will not be validated in PR check Travis jobs until the changes are committed into the repo.

While checking on the implementation, we found that the Travis request API doesn't support triggering builds on PR references.
(Ref: https://github.com/travis-ci/travis-ci/issues/9907, https://travis-ci.community/t/trigger-build-on-pr-via-api/825)

However, as a workaround added below specified commands in the install section of PR-check jobs in travis.yml. However, If a PR contains changes in the travis.yml file, those won't be validated.
```
git fetch origin +refs/pull/${PR_NUMBER}/merge
git checkout -qf FETCH_HEAD
```

**Comment by cwsolee** [comment]
@nickboldt & @ericwill , We keep this PR as draft for u to take a look before we submit, let us know whether you've any comments. Comments from others are welcome too.

**Comment by nickboldt** [comment]
for the secrets/env vars, looks like we have two options:

a) UI - https://travis-ci.com/github/eclipse-che/che-machine-exec/settings
b) CLI - https://docs.travis-ci.com/user/encryption-keys/ (embed secrets in .travis.yaml using Travis gem)

**Comment by Prabhav-Thali** [comment]
@nickboldt As changes from this PR are taken forward in PR #170. Good to close this PR?

**Comment by nickboldt** [comment]
if everything you need is in #170, then yes.

**Comment by Prabhav-Thali** [comment]
@nickboldt I have updated the PR with changes from #170.

**Comment by nickboldt** [comment]
Running https://app.travis-ci.com/github/eclipse-che/che-machine-exec/builds/235860533 because of this merge.
Also expect to see a PR check from https://github.com/eclipse-che/che-machine-exec/pull/185/files but nothing so far

**Comment by nickboldt** [comment]
https://app.travis-ci.com/github/eclipse-che/che-machine-exec/builds/235860766 (PR check) is green. 

Waiting to see what happens with the commit in https://app.travis-ci.com/github/eclipse-che/che-machine-exec/builds/235860533 -- seems to be stuck on the s390x steps

**Comment by nickboldt** [comment]
enabled travis for :next builds:

https://github.com/eclipse-che/che-machine-exec/pull/186

Running in https://app.travis-ci.com/github/eclipse-che/che-machine-exec/builds/235862504

image

**Comment by nickboldt** [comment]
image

**Comment by nickboldt** [comment]
GH action is complete now too -- took 17 mins longer than the Travis build for 3/4 of the same arches.

**Comment by benoitf** [comment]
on https://github.com/eclipse-che/che-machine-exec/runs/3382034579?check_suite_focus=true I see 'wrong credentials'

```
Run body="{
{
  "@type": "error",
  "error_type": "wrong_credentials",
  "error_message": "access denied"
}
```

**Pull request merged**